### PR TITLE
(935) Correct US date formats in submission entries

### PR DESCRIPTION
**Pull request by leeky**
Ingest was previously allowing through dates formatted as mm/dd/yy as
well as those formatted as dd/mm/yyyy. Only the latter is acceptable by
the data warehouse, so we need to correct the existing data.

This was made more complex once I found entries that were in a third
format dd/mm/yy, which make it non-obvious to determine the format the
date fields were ingested as.

Data migration can be run with:
`rails runner db/data_migrate/20190325121200_fix_us_formatted_dates.rb`

### CODE REVIEW

File: spec/support/object_formatter.rb [hunk 1]
```diff
@@ -0,0 +1 @@
+RSpec::Support::ObjectFormatter.default_instance.max_formatted_output_length = 1000
```
Comment (james) [inline spec/support/object_formatter.rb]: This clashes with https://github.com/dxw/DataSubmissionServiceAPI/blob/develop/spec/support/export_spec_expectation_truncations.rb - we should either do it that way, or this way, but not both!

File: spec/models/export/submission_entry_row_spec.rb [hunk 2]
```diff
@@ -62,11 +62,11 @@
       expect(row.formatted_date('21/9/2017')).to eq '2017-09-21'
     end
 
-    it 'handles US-formatted MM/DD/YY date strings' do
-      expect(row.formatted_date('9/10/18')).to eq '2018-09-10'
-      expect(row.formatted_date('8/1/18')).to eq '2018-08-01'
-      expect(row.formatted_date('2/28/17')).to eq '2017-02-28'
-      expect(row.formatted_date('3/28/19')).to eq '2019-03-28'
+    it 'returns the input value for US-formatted MM/DD/YY date strings' do
```
Comment (james) [inline spec/models/export/submission_entry_row_spec.rb]: We've agreed that these specs shouldn't exist any more.

File: db/data_migrate/20190325121200_fix_us_formatted_dates.rb [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+  dates_in_submission = submission_entries.flat_map do |entry|
+    entry.data.select { |k, v| k.match?(/Date/) && v.size > 5 && v.size < 9 }.values
+  end.compact.uniq
+
+  # Detect date format
+  first_component_maybe_day = dates_in_submission.any? { |date| date.split('/')[0].to_i > 12 }
+  second_component_maybe_day = dates_in_submission.any? { |date| date.split('/')[1].to_i > 12 }
+
+  input_format = if first_component_maybe_day && second_component_maybe_day
```
Comment (CristinaRO) [inline db/data_migrate/20190325121200_fix_us_formatted_dates.rb]: I am probably missing something: what happens if neither the first nor the second component are >12? There's no way to distinguish '10/11/18' (US) from '11/10/18' (UK), so shouldn't it say that it's `:unknown`?

File: db/data_migrate/20190325121200_fix_us_formatted_dates.rb [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+puts 'Fixing submission entries...'
+
+affected_submission_ids = SubmissionEntry
+                          .validated
+                          .distinct
+                          .select(:submission_id)
+                          .joins(:submission)
+                          .where(
+                            "(LENGTH(data->>'Customer Invoice Date') IN (6,7,8) " \
```
Comment (CristinaRO) [inline db/data_migrate/20190325121200_fix_us_formatted_dates.rb]: I don't understand the length of this field.

If the date includes the separator `/`, wouldn't the minimum length for `dd/mm/yy` be 8?

Comment (leeky) [reply db/data_migrate/20190325121200_fix_us_formatted_dates.rb]: The date is sometimes not zero-prefixed - so possible values could be `1/1/18`  or `1/12/18` or `12/12/18`.

File: db/data_migrate/20190325121200_fix_us_formatted_dates.rb (hunk 3 above)
Comment (leeky) [reply db/data_migrate/20190325121200_fix_us_formatted_dates.rb]: :+1:

I think when testing this never actually happened (which suprised me!), but will add it back for completeness.

File: db/data_migrate/20190325121200_fix_us_formatted_dates.rb (hunk 4 above)
Comment (CristinaRO) [reply db/data_migrate/20190325121200_fix_us_formatted_dates.rb]: Thanks! I also thought some dates could contain a four-digit year, or was that only at the first stage of their being ingested?

Comment (leeky) [reply db/data_migrate/20190325121200_fix_us_formatted_dates.rb]: If they contain a four-digit year, they are already in the correct format, so we skip those.

**Pull request merged**